sdk/lib/atl/atltrace.h

```diff
@@ -0,0 +1,203 @@
+DEFINE_TRACE_CATEGORY(atlTraceDBClient,   (1 << 8));
+DEFINE_TRACE_CATEGORY(atlTraceDBProvider, (1 << 9));
+DEFINE_TRACE_CATEGORY(atlTraceSnapin,     (1 << 10));
+DEFINE_TRACE_CATEGORY(atlTraceNotImpl,    (1 << 11));
+DEFINE_TRACE_CATEGORY(atlTraceAllocation, (1 << 12));
+#undef DEFINE_TRACE_CATEGORY
+
+struct CTrace
+{
+    enum
+    {
+        DefaultTraceLevel = 0,
+        DisableTracing = 0xFFFFFFFF,
+        EnableAllCategories = 0xFFFFFFFF
+    };
+
+    static UINT GetLevel()      { return s_level; }
+    static UINT GetCategories() { return s_categories; }
+    static void SetLevel(UINT level)           { s_level = level; }
+    static void SetCategories(UINT categories) { s_categories = categories; }
+
+    static bool IsTracingEnabled(UINT category, UINT level)
+    {
+        if (s_level == DisableTracing || s_level < level || !(s_categories & category))
+            return false;
+        return ::IsDebuggerPresent();
+    }
+
+protected:
+    static UINT s_categories;
+    static UINT s_level;
+};
+
+DECLSPEC_SELECTANY UINT CTrace::s_categories = CTrace::EnableAllCategories;
+DECLSPEC_SELECTANY UINT CTrace::s_level      = CTrace::DefaultTraceLevel;
+
+inline VOID __stdcall
+AtlTraceV(_In_opt_z_                     PCSTR   file,
+          _In_                           INT     line,
+          _In_                           UINT    cat,
+          _In_                           UINT    level,
+          _In_z_ _Printf_format_string_  PCSTR   format,
+          _In_                           va_list va)
+{
+    char szBuff[1024];
+    size_t cch;
+
+    if (!CTrace::IsTracingEnabled(cat, level))
+        return;
+
+#ifdef _STRSAFE_H_INCLUDED_
+    StringCchPrintfA(szBuff, _countof(szBuff), "%s(%d) : ", file, line);
+    StringCchLengthA(szBuff, _countof(szBuff), &cch);
+    StringCchVPrintfA(&szBuff[cch], _countof(szBuff) - cch, format, va);
+#else
+    cch = _snprintf(szBuff, _countof(szBuff), "%s(%d) : ", file, line);
+    _vsnprintf(&szBuff[cch], _countof(szBuff) - cch, format, va);
+#endif
+
+    OutputDebugStringA(szBuff);
+}
+
+inline VOID __stdcall
+AtlTraceV(_In_opt_z_                     PCSTR   file,
+          _In_                           INT     line,
+          _In_                           UINT    cat,
+          _In_                           UINT    level,
+          _In_z_ _Printf_format_string_  PCWSTR  format,
+          _In_                           va_list va)
+{
+    WCHAR szBuff[1024];
+    size_t cch;
+
+    if (!CTrace::IsTracingEnabled(cat, level))
+        return;
+
+#ifdef _STRSAFE_H_INCLUDED_
+    StringCchPrintfW(szBuff, _countof(szBuff), L"%hs(%d) : ", file, line);
+    StringCchLengthW(szBuff, _countof(szBuff), &cch);
+    StringCchVPrintfW(&szBuff[cch], _countof(szBuff) - cch, format, va);
+#else
+    cch = _snwprintf(szBuff, _countof(szBuff), L"%hs(%d) : ", file, line);
+    _vsnwprintf(&szBuff[cch], _countof(szBuff) - cch, format, va);
+#endif
+
+    OutputDebugStringW(szBuff);
```

Comment:
If you use `_CrtDbgReportW` there should be enough customization options?

Reply:
OK.